tfhe/src/high_level_api/integers/types/base.rs

```diff
@@ -111,38 +111,76 @@ where
+        P2: IntegerParameter,
         P2::Id: Default,
     {
         crate::high_level_api::global_state::with_internal_keys(|keys| {
             let integer_key = keys.integer_key.pbs_key();
             let current_num_blocks = P::num_blocks();
             let target_num_blocks = P2::num_blocks();
 
-            if target_num_blocks > current_num_blocks {
-                let num_blocks_to_add = target_num_blocks - current_num_blocks;
-                integer_key.extend_radix_with_trivial_zero_blocks_msb_assign(
-                    &mut self.ciphertext,
-                    num_blocks_to_add,
-                );
+            let blocks = if P::InnerCiphertext::IS_SIGNED {
+                if target_num_blocks > current_num_blocks {
+                    let mut ct_as_signed_radix =
+                        SignedRadixCiphertext::from_blocks(self.ciphertext.into_blocks());
+                    let num_blocks_to_add = target_num_blocks - current_num_blocks;
+                    integer_key.extend_radix_with_sign_msb_assign(
+                        &mut ct_as_signed_radix,
+                        num_blocks_to_add,
+                    );
+                    ct_as_signed_radix.blocks
+                } else {
+                    let mut ct_as_unsigned_radix =
+                        RadixCiphertext::from_blocks(self.ciphertext.into_blocks());
+                    let num_blocks_to_remove = current_num_blocks - target_num_blocks;
+                    integer_key.trim_radix_blocks_msb_assign(
+                        &mut ct_as_unsigned_radix,
+                        num_blocks_to_remove,
+                    );
+                    ct_as_unsigned_radix.blocks
+                }
             } else {
-                let num_blocks_to_remove = current_num_blocks - target_num_blocks;
-                integer_key
-                    .trim_radix_blocks_msb_assign(&mut self.ciphertext, num_blocks_to_remove);
-            }
-            GenericInteger::<P2>::new(self.ciphertext, P2::Id::default())
+                let mut ct_as_unsigned_radix =
+                    RadixCiphertext::from_blocks(self.ciphertext.into_blocks());
+                if target_num_blocks > current_num_blocks {
+                    let num_blocks_to_add = target_num_blocks - current_num_blocks;
+                    integer_key.extend_radix_with_trivial_zero_blocks_msb_assign(
+                        &mut ct_as_unsigned_radix,
+                        num_blocks_to_add,
+                    );
+                } else {
+                    let num_blocks_to_remove = current_num_blocks - target_num_blocks;
+                    integer_key.trim_radix_blocks_msb_assign(
+                        &mut ct_as_unsigned_radix,
+                        num_blocks_to_remove,
+                    );
+                }
+                ct_as_unsigned_radix.blocks
+            };
+
+            assert_eq!(blocks.len(), P2::num_blocks());
```

Comment:
error message could be welcome

Reply:
I did not put one because if that assert is hit, that means we (and not the user) did something wrong
but i can add a small one

Reply:
ah right, it never hurts to have a bit of an explicit message (even just for us, this way we can search/grep for the error message)